doc: mgmt: mcumgr: Add details on making handlers

nordicjm opened the pull request:
Adds a guide on how to make out-of-tree MCUmgr function handlers and groups, with an example showing a test implementation.

Fixes #59492

nordicjm committed:
doc: mgmt: mcumgr: Add details on making handlers

Adds a guide on how to make out-of-tree MCUmgr function handlers
and groups, with an example showing a test implementation.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

de-nordic reviewed (dismissed):
Great! Thanks!

fabiobaltieri commented on doc/services/device_mgmt/mcumgr_handlers.rst:
```diff
@@ -0,0 +1,516 @@
+register the transport with MCUmgr so that commands will be sent to it. An example file would
+look similar to:
+
+.. code-block:: c
+
```
Wow that's a lot of code for a document. Thinking, it's going to be hard to keep this up to date, would it be an option to have these live in actual C files in samples or tests and have them build by the CI and included here instead? I think Go has a feature to build code for examples and that's useful to make sure stuff don't go stale, having so much code in document seems like asking for troubles.

@kartben @gmarull

kartben replied:
Thanks for the flag, @fabiobaltieri, and +1.
Is there a reason why this couldn't be an actual sample? Don't get me wrong, I do love the idea of inlining the code snippets in the doc, but this could be done using `.. literalinclude:: ...`, an under-utilized Sphinx feature IMO :) (and it can be pretty flexible too, i.e. only sections of a file can be include, etc.)

keith-zephyr replied:
+1 on using `literalinclude::` pointing to actual code.  The [USB-C power delivery documentation](https://github.com/zephyrproject-rtos/zephyr/blob/3a24476fb70b813cf5ba01853c10f5c7973dbfbc/doc/connectivity/usb/pd/ucds.rst?plain=1#L39) provides some good examples.

nordicjm replied:
The problem I have with making this a sample or a test is for a test, I don't want to test this, this functionality is tested already in other tests, to make all the code for testing this operation is an absolute time sink and doesn't provide anything. As for a sample, I don't really see the point, because no-one would actually use this code as-is, they would look at it to refer to how handlers are implemented but create their own (because to use this handler, you would need a corresponding MCUmgr client which supported it, which they would likely have to go and create themselves). So it can be made a sample if wanted but isn't going to be a useful sample. MCUmgr is a stable API so these APIs are not going to change unless the existing APIs are deprecated or a stable API change request is made

fabiobaltieri replied:
I don't think the idea was tho have it runnable, the idea was to have CI checking that it builds as the rest of the ecosystem moves. But it'd also make it easier to use, one can just copy the files out rather than copypasting the content. How much work is to move these under `tests`, and add the support files to make it a build only test?

nordicjm replied:
Nice, that does work pretty well, have updated it

fabiobaltieri reviewed (changes requested):
(blocking until https://github.com/zephyrproject-rtos/zephyr/pull/59604#discussion_r1267199990 is resolved)

nordicjm committed:
tests: mgmt: mcumgr: Add handler_demo test

Adds a test that tests a build-only configuration of an
application which uses custom handlers is able to build, and
can be referenced in documentation.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

fabiobaltieri reviewed (approved):
Nice!

fabiobaltieri commented:
@kartben anything to add?
